sdk/dotnet/Pulumi/Testing/Mocks.cs

```diff
@@ -0,0 +1,210 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Google.Protobuf.Collections;
+using Google.Protobuf.WellKnownTypes;
+using Pulumi.Serialization;
+using Pulumirpc;
+
+namespace Pulumi.Testing
+{
+    /// <summary>
+    /// Hooks to mock the engine and provide test doubles for offline unit testing of stacks.
+    /// </summary>
+    public interface IMocks
+    {
+        /// <summary>
+        /// Invoked when a new resource is created by the program.
+        /// </summary>
+        /// <param name="resource">Resource instance.</param>
+        /// <param name="type">Resource type name.</param>
+        /// <param name="name">Resource name.</param>
+        /// <param name="inputs">Dictionary of resource input properties.</param>
+        /// <param name="provider">Provider.</param>
+        /// <param name="id">Resource identifier.</param>
+        /// <returns>A tuple of a resource identifier and resource state. State can be either a POCO
+        /// or a dictionary bag.</returns>
+        Task<(string id, object state)> NewResourceAsync(Resource resource, string type, string name,
```

Comment:
I don't think that passing the resource in is particularly safe: the resource won't really be ready until after this call has returned.

Reply:
Indeed, removed.